idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/resolve/ResolveProviderSyntax.kt

```diff
@@ -0,0 +1,84 @@
+package arrow.meta.ide.dsl.resolve
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.phases.ExtensionPhase
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.ResolveScopeProvider
+import com.intellij.psi.search.GlobalSearchScope
+
+/**
+ * [GlobalSearchScope] search for {@code VirtualFile}s in global scope.
+ * [ResolveProviderSyntax] provides APIs to create GlobalSearchScope.
+ */
+interface ResolveProviderSyntax {
+
+
+    /**
+     * registers an [GlobalSearchScope].
+     * One minimal example from [KotlinScriptResolveScopeProvider], may look like this:
+     * ```kotlin
+     * import arrow.meta.ide.IdeMetaPlugin
+     * import arrow.meta.Plugin
+     * import com.intellij.openapi.project.Project
+     * import com.intellij.openapi.vfs.VirtualFile
+     * import com.intellij.psi.PsiManager
+     * import com.intellij.psi.search.GlobalSearchScope
+     * import org.jetbrains.kotlin.idea.KotlinFileType
+     * import org.jetbrains.kotlin.idea.core.script.ScriptConfigurationManager
+     * import org.jetbrains.kotlin.idea.core.script.StandardIdeScriptDefinition
+     * import org.jetbrains.kotlin.psi.KtFile
+     * import org.jetbrains.kotlin.scripting.definitions.ScriptDefinition
+     * import org.jetbrains.kotlin.scripting.definitions.findScriptDefinition
+     * import org.jetbrains.kotlin.scripting.resolve.KotlinScriptDefinitionFromAnnotatedTemplate
+     *
+     * IdeMetaPlugin().addResolveScopeProvider { file: VirtualFile, project: Project ->
+     *  if (file.fileType != KotlinFileType.INSTANCE) return@addResolveScopeProvider null
```

Comment:
Do something with `file.takeIf {it.fileType}?.let { ...}`.

Reply:
something like this:


```
import arrow.meta.ide.IdeMetaPlugin
import arrow.meta.phases.ExtensionPhase
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiManager
import com.intellij.psi.search.GlobalSearchScope
import org.jetbrains.kotlin.idea.KotlinFileType
import org.jetbrains.kotlin.idea.core.script.ScriptConfigurationManager
import org.jetbrains.kotlin.scripting.definitions.ScriptDefinition
import org.jetbrains.kotlin.scripting.definitions.findScriptDefinition
import org.jetbrains.kotlin.scripting.resolve.KotlinScriptDefinitionFromAnnotatedTemplate

val IdeMetaPlugin.scopeProvider: ExtensionPhase
    get() = addResolveScopeProvider { file: VirtualFile, project: Project ->
        file.takeIf { it.fileType == KotlinFileType.INSTANCE }?.let {
            PsiManager.getInstance(project).findFile(it)?.findScriptDefinition()
                    ?.let { scriptDefinition ->
                        when {
                            scriptDefinition is ScriptDefinition.FromConfigurations || scriptDefinition.asLegacyOrNull<KotlinScriptDefinitionFromAnnotatedTemplate>() != null -> {
                                GlobalSearchScope.fileScope(project, file)
                                        .union(ScriptConfigurationManager.getInstance(project).getScriptDependenciesClassFilesScope(file))
                            }
                            else -> null
                        }

                    }
        }
    }
```

Reply:
That looks better. The last `when` is not needed. 
Something like:
```kotlin
file.takeIf {
  it.fileType == KotlinFileType.INSTANCE && 
  PsiManager.getInstance(project)
   .findFile(it)
   ?.findScriptDefinition()
   ?.safeAs<ScriptDefinition.FromConfigurations>()
   ?.asLegacyOrNull<KotlinScriptDefinitionFromAnnotatedTemplate>()
      != null
  }.let { file ->
   GlobalSearchScope
    .fileScope(project, file)
    .union(
      ScriptConfigurationManager
      .getInstance(project)
      .getScriptDependenciesClassFilesScope(file)
     )
}
```